source/fab/errors.py

```diff
@@ -0,0 +1,344 @@
+    instance is requested.
+
+    :param library: the name of the unknown library
+    """
+
+    def __init__(self, library: str) -> None:
+        self.library = library
+        super().__init__(f"unknown library {library}")
+
+
+class FabCommandError(FabError):
+    """An error was encountered running a subcommand.
+
+    Error where a subcommand run by the fab framework returned with a
+    non-zero exit code.  The exit code plus any data from the stdout
+    and stderr streams are retained to allow them to be passed back up
+    the calling stack.
+
+    :param command: the command being run
+    :param code: the command return code from the OS
+    :param output: output stream from the command
+    :param error: error stream from the command
+    """
+
+    def __init__(
+        self,
+        command: str,
+        code: int,
+        output: Union[str, bytes, None],
+        error: Union[str, bytes, None],
+        cwd: Optional[Union[str, Path]] = None,
+    ) -> None:
+        if isinstance(command, list):
+            self.command: str = " ".join(command)
+        else:
+            self.command = str(command)
+        self.code = int(code)
+        self.output = self._decode(output)
+        self.error = self._decode(error)
+        self.cwd = cwd
+        super().__init__(f"command {repr(self.command)} returned {code}")
+
+    def _decode(self, value: Union[str, bytes, None]) -> str:
+        """Convert from bytes to a string as necessary."""
+        if value is None:
+            return ""
+        if isinstance(value, bytes):
+            return value.decode()
+        return value
+
+
+class FabCommandNotFound(FabError):
+    """Target command could not be found by subprocess.
+
+    Error where the target command passed to a subprocess call could
+    not be found.
+
+    :param command: the target command.  For clarity, only the first
+        item is used in the error message but the entire command is
+        preserved for inspection by the caller.
+    """
+
+    def __init__(self, command: str) -> None:
+        self.command = command
+        if isinstance(command, list):
+            self.target: str = command[0]
+        elif isinstance(command, str):
+            self.target = command.split()[0]
+        else:
+            raise ValueError(f"invalid command: {command}")
+
+        super().__init__(f"unable to execute {self.target}")
+
+
+class FabMultiCommandError(FabError):
+    """Unpack multiple exceptions into a single one.
+
+    Error which combines all potential exceptions raised by a
+    multiprocessing section into a single exception class for
+    subsequent inspection.
+
+    This feature is is required because versions of python prior to
+    3.11 do not support ExceptionGroups.
+
+    :param errors: a list ot exceptions
+    :param label: an identifier for the multiprocessing section
+    """
+
+    def __init__(
+        self, errors: List[Union[str, Exception]], label: Optional[str] = None
```

Comment:
If this exception consolidates exceptions, why is it accepting strings?

If that is useful, should it create an exception to hold the string so `self.errors` holds only exceptions?